core/optaplanner-constraint-streams-common/src/test/java/org/optaplanner/constraint/streams/common/AbstractConstraintStreamScoreDirectorSemanticsTest.java

```diff
@@ -0,0 +1,100 @@
+    }
+
+    protected <Solution_> InnerScoreDirectorFactory<Solution_, SimpleScore>
+            buildInnerScoreDirectorFactory(SolutionDescriptor<Solution_> solutionDescriptor,
+                    Class<? extends ConstraintProvider> constraintProviderClass) {
+        ScoreDirectorFactoryConfig scoreDirectorFactoryConfig = new ScoreDirectorFactoryConfig()
+                .withConstraintProviderClass(constraintProviderClass)
+                .withConstraintStreamImplType(getConstraintStreamImplType());
+        ScoreDirectorFactoryFactory<Solution_, SimpleScore> scoreDirectorFactoryFactory =
+                new ScoreDirectorFactoryFactory<>(scoreDirectorFactoryConfig);
+        return scoreDirectorFactoryFactory.buildScoreDirectorFactory(getClass().getClassLoader(),
+                EnvironmentMode.REPRODUCIBLE, solutionDescriptor);
+    }
+
+    @Test
+    void polymorphismNullityFilter() {
+        SolutionDescriptor<TestdataSolution> solutionDescriptor =
+                TestdataSolution.buildSolutionDescriptor();
+        InnerScoreDirectorFactory<TestdataSolution, SimpleScore> scoreDirectorFactory =
+                buildInnerScoreDirectorFactory(solutionDescriptor, NullityFilterConstraintProvider.class);
+
+        TestdataSolution solution = TestdataSolution.generateSolution(1, 2);
+        TestdataEntity uninitializedEntity = solution.getEntityList().get(0);
+        uninitializedEntity.setValue(null);
+        try (InnerScoreDirector<TestdataSolution, SimpleScore> scoreDirector =
+                scoreDirectorFactory.buildScoreDirector(false, true)) {
+            // One of the entities has a null variable, it will therefore be uninitialized; standard forEach() behavior.
+            scoreDirector.setWorkingSolution(solution);
+            Assertions.assertThat(scoreDirector.calculateScore())
+                    .isEqualTo(SimpleScore.ofUninitialized(-1, -2)); // One value, one initialized entity.
+
+            scoreDirector.beforeVariableChanged(uninitializedEntity, "value");
+            uninitializedEntity.setValue(new TestdataValue());
+            scoreDirector.afterVariableChanged(uninitializedEntity, "value");
+
+            // The entity is no longer null, and it will no longer be skipped.
+            scoreDirector.triggerVariableListeners();
+            Assertions.assertThat(scoreDirector.calculateScore())
+                    .isEqualTo(SimpleScore.of(-3)); // One value, two initialized entities.
+        }
+    }
+
+    public static final class NullityFilterConstraintProvider implements ConstraintProvider {
+
+        @Override
+        public Constraint[] defineConstraints(ConstraintFactory constraintFactory) {
+            return new Constraint[] {
+                    constraintFactory.forEach(Object.class)
+                            .penalize("Penalize any entity with null variable", SimpleScore.ONE)
```

Comment:
And this is what it makes possible.
Entities that match will now have their nullity filters checked.